src/Microsoft.AspNet.Mvc.Core/Internal/Routing/LinkGenerationMatch.cs

```diff
@@ -10,14 +13,42 @@ public struct LinkGenerationMatch
         private readonly bool _isFallbackMatch;
         private readonly AttributeRouteLinkGenerationEntry _entry;
 
-        public LinkGenerationMatch(AttributeRouteLinkGenerationEntry entry, bool isFallbackMatch)
+        public LinkGenerationMatch(
+            AttributeRouteLinkGenerationEntry entry,
+            bool isFallbackMatch,
+            VirtualPathContext pathContext)
         {
             _entry = entry;
             _isFallbackMatch = isFallbackMatch;
+            RouteValues = pathContext.Values;
+            AmbientValues = pathContext.AmbientValues;
         }
 
         public AttributeRouteLinkGenerationEntry Entry { get { return _entry; } }
 
         public bool IsFallbackMatch { get { return _isFallbackMatch; } }
+
+        public IDictionary<string, object> RouteValues { get; }
+
+        public IDictionary<string, object> AmbientValues { get; }
+
+        public int ParametersWithValues
+        {
+            get
+            {
+                var t = this;
```

Comment:
If you're complaining about aliasing this that's because this isn't available for a lambda inside a struct. It'll get removed when I get rid of the Linq anyway.